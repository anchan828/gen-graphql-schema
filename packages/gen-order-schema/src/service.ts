import {
  buildASTSchema,
  DirectiveNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  parse,
  printSchema,
} from 'graphql';
import { DEFAULT_OPTIONS } from './constants';
import { GenOrderTypesOptions } from './options';
import {
  getDirectives,
  getFieldDefinitions,
  getFieldTypeName,
  getObjectTypeDefinitions,
  isBasicType,
  isEnumType,
} from './utils';

export class GenOrderTypesService {
  genOrderTypes(): string {
    if (!this.hasOrderByDirective()) {
      return printSchema(buildASTSchema(this.documentNode));
    }

    const fields = this.getOrderByFieldDefinitions();

    for (const field of fields) {
      const enumTypeDefinition = this.genOrderByEnumTypeDefinition(field);
      if (!enumTypeDefinition) {
        continue;
      }

      this.appendDefinitionToDocumentNode(enumTypeDefinition);
      this.appendOrderByArgumentToFieldNode(field, enumTypeDefinition);
      this.removeOrderByDirective(field);
      this.removeOrderByIgnoreDirective(field);
    }

    return printSchema(buildASTSchema(this.documentNode));
  }
  private hasOrderByDirective(): boolean {
    const definitions = getObjectTypeDefinitions(this.documentNode);
    for (const definition of definitions) {
      const fields = getFieldDefinitions(definition);

      for (const field of fields) {
        const directives = getDirectives(field);
        if (
          directives.find(
            (d: DirectiveNode) =>
              d.name.value === this.options.orderByDirectiveName,
          )
        ) {
          return true;
        }
      }
    }

    return false;
  }
  private getOrderByFieldDefinitions(): FieldDefinitionNode[] {
    const results: FieldDefinitionNode[] = [];
    const definitions = getObjectTypeDefinitions(this.documentNode);
    for (const definition of definitions) {
      const fields = getFieldDefinitions(definition);

      for (const field of fields) {
        const directives = getDirectives(field);
        if (
          directives.find(
            (d: DirectiveNode) =>
              d.name.value === this.options.orderByDirectiveName,
          )
        ) {
          const { name, isList } = getFieldTypeName(field);
          if (!isList) {
            console.warn(
              `Found order directive, but type of ${
                field.name.value
              } was not ListType. So skip.`,
            );
          } else if (isBasicType(name)) {
            console.warn(
              `Found order directive, but type of ${
                field.name.value
              } was basic types. So skip.`,
            );
          } else {
            results.push(field);
          }
        }
      }
    }
    return results;
  }
  private getObjectTypeDefinition(
    field: FieldDefinitionNode,
  ): ObjectTypeDefinitionNode | undefined {
    const { name } = getFieldTypeName(field);

    return getObjectTypeDefinitions(this.documentNode).find(
      definition => definition.name.value === name,
    );
  }

  private hasOrderByIgnoreDirective(field: FieldDefinitionNode): boolean {
    const directives = getDirectives(field);
    if (
      directives.find(
        (d: DirectiveNode) =>
          d.name.value === this.options.orderByIgnoreDirectiveName,
      )
    ) {
      return true;
    }

    return false;
  }
  private getOrderableFieldNames(
    definition: ObjectTypeDefinitionNode,
  ): string[] {
    const orderableFieldNames: string[] = [];
    for (const field of getFieldDefinitions(definition)) {
      const { name } = getFieldTypeName(field);

      if (
        (isBasicType(name) ||
          isEnumType(this.documentNode, name) ||
          this.options.supportOrderableTypes!.includes(name)) &&
        !this.hasOrderByIgnoreDirective(field)
      ) {
        orderableFieldNames.push(field.name.value);
      }
    }

    return orderableFieldNames;
  }

  private genOrderByEnumTypeDefinition(
    field: FieldDefinitionNode,
  ): EnumTypeDefinitionNode | undefined {
    const type = this.getObjectTypeDefinition(field);
    if (!type) {
      return;
    }
    const { name } = getFieldTypeName(field);
    const orderableFieldNames = this.getOrderableFieldNames(type);
    return {
      kind: 'EnumTypeDefinition',
      name: {
        kind: 'Name',
        value: `${name}${this.options.orderEnumTypeSuffix}`,
      },
      directives: [],
      values: [
        ...Array.prototype.concat.apply(
          [],
          orderableFieldNames.map(orderableFieldName => [
            {
              kind: 'EnumValueDefinition',
              name: {
                kind: 'Name',
                value: `${orderableFieldName}_ASC`,
              },
              directives: [],
            },
            {
              kind: 'EnumValueDefinition',
              name: {
                kind: 'Name',
                value: `${orderableFieldName}_DESC`,
              },
              directives: [],
            },
          ]),
        ),
      ],
    };
  }
  private appendDefinitionToDocumentNode(
    enumTypeDefinition: EnumTypeDefinitionNode,
  ): void {
    Reflect.set(this.documentNode, 'definitions', [
      ...this.documentNode.definitions,
      enumTypeDefinition,
    ]);
  }
  private appendOrderByArgumentToFieldNode(
    field: FieldDefinitionNode,
    enumTypeDefinition: EnumTypeDefinitionNode,
  ): void {
    Reflect.set(field, 'arguments', [
      ...(field.arguments || []),
      this.genOrderByArgument(enumTypeDefinition),
    ]);
  }

  private genOrderByArgument(
    enumType: EnumTypeDefinitionNode,
  ): InputValueDefinitionNode {
    let type: NamedTypeNode | ListTypeNode = {
      kind: 'NamedType',
      name: {
        kind: 'Name',
        value: `${enumType.name.value}`,
      },
    };

    if (this.options.orderByArgumentTypeIsList) {
      type = {
        kind: 'ListType',
        type,
      };
    }

    return {
      kind: 'InputValueDefinition',
      name: {
        kind: 'Name',
        value: this.options.orderByArgumentName!,
      },
      type,
      directives: [],
    };
  }
  private removeOrderByDirective(field: FieldDefinitionNode): void {
    const directives = getDirectives(field);
    Reflect.set(
      field,
      'directives',
      directives.filter(
        directive => directive.name.value !== this.options.orderByDirectiveName,
      ),
    );
  }
  private removeOrderByIgnoreDirective(field: FieldDefinitionNode): void {
    const type = this.getObjectTypeDefinition(field);
    if (!type) {
      return;
    }

    for (const f of getFieldDefinitions(type)) {
      if (!this.hasOrderByIgnoreDirective(f)) {
        continue;
      }
      Reflect.set(
        f,
        'directives',
        getDirectives(f).filter(
          directive =>
            directive.name.value !== this.options.orderByIgnoreDirectiveName,
        ),
      );
    }
  }
  private readonly documentNode: DocumentNode;
  constructor(
    readonly types: string | DocumentNode,
    private readonly options: GenOrderTypesOptions = {},
  ) {
    this.options = Object.assign({}, DEFAULT_OPTIONS, options);
    this.documentNode = typeof types === 'string' ? parse(types) : types;
  }
}
