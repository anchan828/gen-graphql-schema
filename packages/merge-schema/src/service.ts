import {
  getObjectOrUnionTypeDefinitions,
  getObjectTypeDefinitionsFromUnion,
  isUnionType,
} from '@anchan828/gen-graphql-schema-common';
import {
  DocumentNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  parse,
  StringValueNode,
} from 'graphql';
type TypeName = string;
type Description = string;
export class MergeSchemaService {
  private descriptionMap: Map<TypeName, Description> = new Map<
    TypeName,
    Description
  >();

  private interfaceMap: Map<TypeName, InterfaceTypeDefinitionNode[]> = new Map<
    TypeName,
    InterfaceTypeDefinitionNode[]
  >();

  public cache(type: string | DocumentNode): void {
    if (typeof type === 'string') {
      type = parse(type);
    }
    this.cacheDescriptions(type);
    this.cacheInterfaces(type);
  }
  public apply(type: DocumentNode): void {
    this.applyDescriptions(type);
    this.applyInterfaces(type);
  }
  private cacheInterfaces(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (!Array.isArray(definition.interfaces)) {
        continue;
      }
      if (!this.interfaceMap.has(definition.name.value)) {
        this.interfaceMap.set(definition.name.value, []);
      }

      const cachedInterfaces = this.interfaceMap.get(definition.name.value)!;

      for (const interfaceDefinition of definition.interfaces) {
        if (
          !cachedInterfaces.find(
            ci => interfaceDefinition.name.value === ci.name.value,
          )
        ) {
          cachedInterfaces.push(interfaceDefinition);
        }
      }
      this.interfaceMap.set(definition.name.value, cachedInterfaces);
    }
  }
  private cacheDescriptions(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (!(definition.description && definition.description.value)) {
        continue;
      }

      let description = definition.description.value;

      if (this.descriptionMap.has(definition.name.value)) {
        description = [
          this.descriptionMap.get(definition.name.value),
          definition.name.value,
        ].join('\n');
      }

      this.descriptionMap.set(definition.name.value, description);
    }
  }

  public applyDescriptions(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (this.descriptionMap.has(definition.name.value)) {
        Reflect.set(definition, 'description', {
          kind: 'StringValue',
          value: this.descriptionMap.get(definition.name.value),
        } as StringValueNode);
      }
    }
  }
  public applyInterfaces(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (this.interfaceMap.has(definition.name.value)) {
        const interfaces = Object.assign(
          [],
          definition.interfaces || [],
        ) as InterfaceTypeDefinitionNode[];
        const cachedInterfaces = this.interfaceMap.get(definition.name.value);
        if (!Array.isArray(cachedInterfaces)) {
          continue;
        }
        for (const cachedInterface of cachedInterfaces) {
          if (
            !interfaces.find(
              (i: InterfaceTypeDefinitionNode) =>
                i.name.value === cachedInterface.name.value,
            )
          ) {
            interfaces.push(cachedInterface);
          }
        }

        Reflect.set(definition, 'interfaces', interfaces);
      }
    }
  }

  private getAllObjectDefinitionsByDocumentNode(
    documentNode: DocumentNode,
  ): ObjectTypeDefinitionNode[] {
    const definitions: ObjectTypeDefinitionNode[] = [];

    for (const objOrUnion of getObjectOrUnionTypeDefinitions(documentNode)) {
      if (isUnionType(objOrUnion)) {
        definitions.push(
          ...getObjectTypeDefinitionsFromUnion(documentNode, objOrUnion),
        );
      } else {
        definitions.push(objOrUnion);
      }
    }
    const cached: Set<string> = new Set<string>();
    return definitions.filter(def => {
      if (!cached.has(def.name.value)) {
        cached.add(def.name.value);
        return true;
      } else {
        return false;
      }
    });
  }
}
