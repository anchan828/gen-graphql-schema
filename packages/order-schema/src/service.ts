import {
  getDirectives,
  getFieldDefinitions,
  getFieldDefinitionsByDirective,
  getFieldTypeName,
  getObjectTypeDefinition,
  getObjectTypeDefinitions,
  isBasicType,
  isEnumType,
} from '@anchan828/gen-graphql-schema-common';
import * as deepmerge from 'deepmerge';
import {
  buildASTSchema,
  DefinitionNode,
  DirectiveNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  parse,
  printSchema,
} from 'graphql';
import { DEFAULT_OPTIONS } from './constants';
import { GenOrderTypesOptions } from './options';
export class GenOrderTypesService {
  genOrderTypes(): string {
    if (!this.hasOrderByDirective()) {
      return printSchema(buildASTSchema(this.documentNode));
    }

    const fields = getFieldDefinitionsByDirective(
      this.documentNode,
      this.options.orderByDirective!.name!,
    );
    this.appendDefinitionToDocumentNode(this.genOrderDirectionEnumDefinition());
    for (const field of fields) {
      const sortEnumType = this.genSortEnumTypeDefinition(field);
      const orderEnumType = this.genOrderObjectTypeDefinition(field);
      if (!sortEnumType || !orderEnumType) {
        continue;
      }

      this.appendDefinitionToDocumentNode(sortEnumType, orderEnumType);
      this.appendOrderByArgumentToFieldNode(field, orderEnumType);

      this.removeOrderByDirective(field);
      this.removeOrderByIgnoreDirective(field);
    }

    return printSchema(buildASTSchema(this.documentNode));
  }
  private hasOrderByDirective(): boolean {
    const definitions = getObjectTypeDefinitions(this.documentNode);
    for (const definition of definitions) {
      const fields = getFieldDefinitions(definition);

      for (const field of fields) {
        const directives = getDirectives(field);
        if (
          directives.find(
            (d: DirectiveNode) =>
              d.name.value === this.options.orderByDirective!.name,
          )
        ) {
          return true;
        }
      }
    }

    return false;
  }

  private hasOrderByIgnoreDirective(field: FieldDefinitionNode): boolean {
    const directives = getDirectives(field);
    if (
      directives.find(
        (d: DirectiveNode) =>
          d.name.value === this.options.orderByIgnoreDirective!.name,
      )
    ) {
      return true;
    }

    return false;
  }
  private getOrderableFieldNames(
    definition: ObjectTypeDefinitionNode,
  ): string[] {
    const orderableFieldNames: string[] = [];
    for (const field of getFieldDefinitions(definition)) {
      const { name } = getFieldTypeName(field);

      if (
        (isBasicType(name) ||
          isEnumType(this.documentNode, name) ||
          this.options.supportOrderableTypes!.includes(name)) &&
        !this.hasOrderByIgnoreDirective(field)
      ) {
        orderableFieldNames.push(field.name.value);
      }
    }

    return orderableFieldNames;
  }

  private genOrderDirectionEnumDefinition(): EnumTypeDefinitionNode {
    const orderDirectionOptions = this.options!.orderDirection!;
    return {
      kind: 'EnumTypeDefinition',
      name: {
        kind: 'Name',
        value: orderDirectionOptions.typeName!,
      },
      directives: [],
      values: [
        {
          kind: 'EnumValueDefinition',
          name: {
            kind: 'Name',
            value: orderDirectionOptions.ascName!,
          },
          directives: [],
        },
        {
          kind: 'EnumValueDefinition',
          name: {
            kind: 'Name',
            value: orderDirectionOptions.descName!,
          },
          directives: [],
        },
      ],
    };
  }

  private genOrderObjectTypeDefinition(
    field: FieldDefinitionNode,
  ): ObjectTypeDefinitionNode | undefined {
    const type = getObjectTypeDefinition(this.documentNode, field);
    if (!type) {
      return;
    }
    const { name } = getFieldTypeName(field);
    const orderTypeOptions = this.options.orderType!;
    const sortEnumOptions = this.options.sortEnum!;
    const orderDirectionOptions = this.options!.orderDirection!;
    return {
      kind: 'ObjectTypeDefinition',
      name: {
        kind: 'Name',
        value: `${orderTypeOptions.prefix}${name}${orderTypeOptions.suffix}`,
      },
      directives: [],
      fields: [
        {
          kind: 'FieldDefinition',
          name: {
            kind: 'Name',
            value: orderTypeOptions.sortName,
          },
          type: {
            kind: 'NamedType',
            name: {
              kind: 'Name',
              value: `${sortEnumOptions.prefix}${name}${
                sortEnumOptions.suffix
              }`,
            },
          },
        },
        {
          kind: 'FieldDefinition',
          name: {
            kind: 'Name',
            value: orderTypeOptions.directionName,
          },
          type: {
            kind: 'NamedType',
            name: {
              kind: 'Name',
              value: orderDirectionOptions.typeName!,
            },
          },
        },
      ],
    } as ObjectTypeDefinitionNode;
  }
  private genSortEnumTypeDefinition(
    field: FieldDefinitionNode,
  ): EnumTypeDefinitionNode | undefined {
    const type = getObjectTypeDefinition(this.documentNode, field);
    if (!type) {
      return;
    }
    const { name } = getFieldTypeName(field);
    const orderableFieldNames = this.getOrderableFieldNames(type);
    const sortEnumOptions = this.options!.sortEnum!;
    return {
      kind: 'EnumTypeDefinition',
      name: {
        kind: 'Name',
        value: `${sortEnumOptions.prefix}${name}${sortEnumOptions.suffix}`,
      },
      directives: [],
      values: orderableFieldNames.map(orderableFieldName => ({
        kind: 'EnumValueDefinition',
        name: {
          kind: 'Name',
          value: `${orderableFieldName.toUpperCase()}`,
        },
        directives: [],
      })) as EnumValueDefinitionNode[],
    };
  }
  private appendDefinitionToDocumentNode(
    ...definitions: DefinitionNode[]
  ): void {
    Reflect.set(this.documentNode, 'definitions', [
      ...this.documentNode.definitions,
      ...definitions,
    ]);
  }
  private appendOrderByArgumentToFieldNode(
    field: FieldDefinitionNode,
    orderTypeDefinition: ObjectTypeDefinitionNode,
  ): void {
    Reflect.set(field, 'arguments', [
      ...(field.arguments || []),
      this.genOrderByArgument(orderTypeDefinition),
    ]);
  }

  private genOrderByArgument(
    orderTypeDefinition: ObjectTypeDefinitionNode,
  ): FieldDefinitionNode {
    let type: NamedTypeNode | ListTypeNode = {
      kind: 'NamedType',
      name: {
        kind: 'Name',
        value: `${orderTypeDefinition.name.value}`,
      },
    };
    const orderByArgumentOptions = this.options!.orderByArgument!;
    if (orderByArgumentOptions.isList) {
      type = {
        kind: 'ListType',
        type,
      };
    }

    return {
      kind: 'FieldDefinition',
      name: {
        kind: 'Name',
        value: orderByArgumentOptions.name!,
      },
      type,
      directives: [],
    };
  }

  private removeOrderByDirective(field: FieldDefinitionNode): void {
    const directives = getDirectives(field);
    Reflect.set(
      field,
      'directives',
      directives.filter(
        directive =>
          directive.name.value !== this.options.orderByDirective!.name,
      ),
    );
  }
  private removeOrderByIgnoreDirective(field: FieldDefinitionNode): void {
    const type = getObjectTypeDefinition(this.documentNode, field);
    if (!type) {
      return;
    }

    for (const f of getFieldDefinitions(type)) {
      if (!this.hasOrderByIgnoreDirective(f)) {
        continue;
      }
      Reflect.set(
        f,
        'directives',
        getDirectives(f).filter(
          directive =>
            directive.name.value !== this.options.orderByIgnoreDirective!.name,
        ),
      );
    }
  }
  private readonly documentNode: DocumentNode;
  constructor(
    readonly types: string | DocumentNode,
    private readonly options: GenOrderTypesOptions = {},
  ) {
    this.options = deepmerge(DEFAULT_OPTIONS, options);
    this.documentNode = typeof types === 'string' ? parse(types) : types;
  }
}
