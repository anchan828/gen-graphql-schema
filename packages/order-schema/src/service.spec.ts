import { buildASTSchema, parse, printSchema } from "graphql";
import { GenOrderTypesService } from "./service";

describe("GenOrderTypesService", () => {
  it("should return same schema when no orderBy directive", () => {
    const types = `type Test { id: ID }`;
    expect(printSchema(buildASTSchema(new GenOrderTypesService(types).genOrderTypes()))).toEqual(
      printSchema(buildASTSchema(parse(types))),
    );
  });

  it("should return added Order enum when has orderBy directive", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenOrderTypesService(
            [`type Test { id: ID }`, `type Query { tests: [Test] @orderBy}`].join(`\n`),
          ).genOrderTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should return added Order enum when change options", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenOrderTypesService(
            [
              `scalar Date`,
              `type Test { id: ID, name: String @ignore, date: Date }`,
              `type Query { tests: [Test] @hoge}`,
            ].join(`\n`),
            {
              orderByDirective: {
                name: "hoge",
              },
              orderByIgnoreDirective: {
                name: "ignore",
              },
              orderByArgument: {
                name: "arg",
                isList: false,
              },
              orderType: {
                suffix: "Changed",
              },
              supportOrderableTypes: ["Date"],
            },
          ).genOrderTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should return added Order enum", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenOrderTypesService(
            [`type Test { id: ID, name: String @ignore }`, `type Query { tests: [[Test!]]! @hoge}`].join(`\n`),
            {
              orderByDirective: {
                name: "hoge",
              },
              orderByIgnoreDirective: {
                name: "ignore",
              },
              orderByArgument: {
                name: "arg",
              },
              orderType: {
                prefix: "Prefix",
                suffix: "Changed",
              },
              orderDirection: {
                ascName: "TOP",
                descName: "BOTTOM",
                typeName: "Position",
              },
              orderFieldEnum: {
                prefix: "Prefix",
                suffix: "Suffix",
              },
            },
          ).genOrderTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should support union type", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenOrderTypesService(
            [
              `type Test1 { id: ID, name: String! }`,
              `type Test2 { id: ID, age: Int! }`,
              `union Test = Test1 | Test2`,
              `type Query { tests: [Test] @orderBy}`,
            ].join(`\n`),
          ).genOrderTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should deep where", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenOrderTypesService(
            [
              `
type TestPosition {
  column: Int!
  line: Int!
  prev: TestPosition @orderBy_nested
  next: TestPosition 
}
type TestSubPosition {
  column: Int!
  line: Int!
}         
type Test {
  id: ID!
  position: TestPosition! @orderBy_nested
}`,
              `type Query { tests: [Test] @orderBy}`,
            ].join(`\n`),
          ).genOrderTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });
});
