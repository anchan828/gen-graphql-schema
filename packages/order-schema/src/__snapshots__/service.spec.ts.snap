// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenOrderTypesService should deep where 1`] = `
"type TestPosition {
  column: Int!
  line: Int!
  prev: TestPosition
  next: TestPosition
}

type TestSubPosition {
  column: Int!
  line: Int!
}

type Test {
  id: ID!
  position: TestPosition!
}

type Query {
  tests(orderBy: TestOrder): [Test]
}

\\"\\"\\"Sort the results in ascending or descending order\\"\\"\\"
enum OrderDirection {
  \\"\\"\\"Sort the results in ascending order\\"\\"\\"
  ASC

  \\"\\"\\"Sort the results in descending order\\"\\"\\"
  DESC
}

\\"\\"\\"Ordering options for TestPosition\\"\\"\\"
input TestPositionOrder {
  \\"\\"\\"Order TestPosition by column\\"\\"\\"
  column: OrderDirection

  \\"\\"\\"Order TestPosition by line\\"\\"\\"
  line: OrderDirection

  \\"\\"\\"Order TestPosition by prev\\"\\"\\"
  prev: TestPositionOrder
}

\\"\\"\\"Ordering options for Test\\"\\"\\"
input TestOrder {
  \\"\\"\\"Order Test by id\\"\\"\\"
  id: OrderDirection

  \\"\\"\\"Order Test by position\\"\\"\\"
  position: TestPositionOrder
}
"
`;

exports[`GenOrderTypesService should return added Order enum 1`] = `
"type Test {
  id: ID
  name: String
}

type Query {
  tests(arg: PrefixTestChanged): [[Test!]]!
}

\\"\\"\\"Sort the results in ascending or descending order\\"\\"\\"
enum Position {
  \\"\\"\\"Sort the results in ascending order\\"\\"\\"
  TOP

  \\"\\"\\"Sort the results in descending order\\"\\"\\"
  BOTTOM
}

\\"\\"\\"Ordering options for Test\\"\\"\\"
input PrefixTestChanged {
  \\"\\"\\"Order Test by id\\"\\"\\"
  id: Position
}
"
`;

exports[`GenOrderTypesService should return added Order enum when change options 1`] = `
"scalar Date

type Test {
  id: ID
  name: String
  date: Date
}

type Query {
  tests(arg: TestChanged): [Test]
}

\\"\\"\\"Sort the results in ascending or descending order\\"\\"\\"
enum OrderDirection {
  \\"\\"\\"Sort the results in ascending order\\"\\"\\"
  ASC

  \\"\\"\\"Sort the results in descending order\\"\\"\\"
  DESC
}

\\"\\"\\"Ordering options for Test\\"\\"\\"
input TestChanged {
  \\"\\"\\"Order Test by id\\"\\"\\"
  id: OrderDirection

  \\"\\"\\"Order Test by date\\"\\"\\"
  date: OrderDirection
}
"
`;

exports[`GenOrderTypesService should return added Order enum when has orderBy directive 1`] = `
"type Test {
  id: ID
}

type Query {
  tests(orderBy: TestOrder): [Test]
}

\\"\\"\\"Sort the results in ascending or descending order\\"\\"\\"
enum OrderDirection {
  \\"\\"\\"Sort the results in ascending order\\"\\"\\"
  ASC

  \\"\\"\\"Sort the results in descending order\\"\\"\\"
  DESC
}

\\"\\"\\"Ordering options for Test\\"\\"\\"
input TestOrder {
  \\"\\"\\"Order Test by id\\"\\"\\"
  id: OrderDirection
}
"
`;

exports[`GenOrderTypesService should support union type 1`] = `
"type Test1 {
  id: ID
  name: String!
}

type Test2 {
  id: ID
  age: Int!
}

union Test = Test1 | Test2

type Query {
  tests(orderBy: TestOrder): [Test]
}

\\"\\"\\"Sort the results in ascending or descending order\\"\\"\\"
enum OrderDirection {
  \\"\\"\\"Sort the results in ascending order\\"\\"\\"
  ASC

  \\"\\"\\"Sort the results in descending order\\"\\"\\"
  DESC
}

\\"\\"\\"Ordering options for Test\\"\\"\\"
input TestOrder {
  \\"\\"\\"Order Test by id\\"\\"\\"
  id: OrderDirection

  \\"\\"\\"Order Test by name\\"\\"\\"
  name: OrderDirection

  \\"\\"\\"Order Test by age\\"\\"\\"
  age: OrderDirection
}
"
`;
